In [29]:
import pandas as pd
import numpy as np
import re
import os

In [30]:
# Đường dẫn thư mục cleaned trong thư mục data
cleaned_data_path = './data/cleaned/'
csv_file_path = os.path.join(cleaned_data_path, 'anime_data_cleaned.csv')

In [31]:
data = pd.read_csv('./data/merge/anime_data.csv', encoding='utf-8')

In [32]:
data.head()

title  \
0                           5-toubun no Hanayome*   
1                Overlord Movie 3: Sei Oukoku-hen   
2             Digimon Adventure 02: The Beginning   
3  Tengen Toppa Gurren Lagann Movie 1: Gurren-hen   
4                     Yowamushi Pedal: Spare Bike   

                                                link  \
0  https://animevietsub.page/phim/5-toubun-no-han...   
1  https://animevietsub.page/phim/overlord-movie-...   
2  https://animevietsub.page/phim/digimon-adventu...   
3  https://animevietsub.page/phim/tengen-toppa-gu...   
4  https://animevietsub.page/phim/yowamushi-pedal...   

                                               image episode  \
0  https://cdn.animevietsub.page/data/poster/2024...     NaN   
1  https://cdn.animevietsub.page/data/poster/2024...     NaN   
2  https://cdn.animevietsub.page/data/poster/2024...     NaN   
3  https://cdn.animevietsub.page/data/poster/2024...     NaN   
4  https://cdn.animevietsub.page/data/poster/2024...     NaN   

                                             rating    views quality  year  \
0  9.6 trong số 10 dựa trên 301 thành viên đánh giá  235,236      HD  2024   
1  9.4 trong số 10 dựa trên 115 thành viên đánh giá  687,666     CAM  2024   
2    7.3 trong số 10 dựa trên 3 thành viên đánh giá   27,447      HD  2023   
3      9 trong số 10 dựa trên 9 thành viên đánh giá   37,285     FHD  2008   
4   7.5 trong số 10 dựa trên 11 thành viên đánh giá   15,265     FHD  2016   

                                      genres  \
0    Shounen, School, Romance, Harem, Comedy   
1          Fantasy, Adventure, Magic, Action   
2  Fantasy, Drama, Comedy, Adventure, Action   
3                             Sci-Fi, Action   
4             Shounen, Sports, Drama, Action   

                                         description   season     type  
0  Chuyến đi tuần trăng mật của Fuutarou Uesugi v...  Mùa thu  phim lẻ  
1  Vương quốc thiêng liêng đã trải qua nhiều năm ...  Mùa thu  phim lẻ  
2  Hai năm sau trận chiến với Eosmon, Davis Motom...  Mùa thu  phim lẻ  
3  Studio hoạt hình Gainax đã giới thiệu một tran...  Mùa thu  phim lẻ  
4  Anime chuyển thể từ manga ngoại truyện cùng tê...  Mùa thu  phim lẻ

In [33]:
data.tail()

title                                               link  \
5384    Lớp Học Ám Sát  https://animevietsub.page/phim/lop-hoc-am-sat-...   
5385  Lớp Học Ám Sát 2  https://animevietsub.page/phim/lop-hoc-am-sat-...   
5386               Air           https://animevietsub.page/phim/air-a186/   
5387           Amnesia    https://animevietsub.page/phim/amnesia-i3-a149/   
5388              Ajin       https://animevietsub.page/phim/ajin-r1-a114/   

                                                  image    episode  \
5384  https://cdn.animevietsub.page/data/poster/2017...  HOÀN\nTẤT   
5385  https://cdn.animevietsub.page/data/poster/2017...  HOÀN\nTẤT   
5386  https://cdn.animevietsub.page/data/poster/2017...  HOÀN\nTẤT   
5387  https://cdn.animevietsub.page/data/poster/2017...  HOÀN\nTẤT   
5388  https://cdn.animevietsub.page/data/poster/2017...  HOÀN\nTẤT   

                                                rating      views quality  \
5384  9.6 trong số 10 dựa trên 191 thành viên đánh giá  2,317,239      HD   
5385  9.6 trong số 10 dựa trên 171 thành viên đánh giá  1,731,882      BD   
5386   8.8 trong số 10 dựa trên 21 thành viên đánh giá     76,301      BD   
5387   8.5 trong số 10 dựa trên 25 thành viên đánh giá    124,034      HD   
5388   8.2 trong số 10 dựa trên 14 thành viên đánh giá    282,158      BD   

      year                                         genres  \
5384  2015                Shounen, School, Comedy, Action   
5385  2016                Shounen, School, Comedy, Action   
5386  2005    Romance, Slice of Life, Supernatural, Drama   
5387  2013               Romance, Fantasy, Mystery, Josei   
5388  2016  Seinen, Supernatural, Horror, Mystery, Action   

                                            description    season     type  
5384  Các học sinh lớp 3-E có 1 nhiệm vụ, đó là ám s...  Mùa đông  phim bộ  
5385                     Phần 2 của Ansatsu Kyoushitsu.  Mùa đông  phim bộ  
5386  \n\nMùa hè thứ 1000…Kunisaki Yukito, một khách...  Mùa đông  phim bộ  
5387  Câu chuyện diễn ra ở một thị trấn hư cấu, tron...  Mùa đông  phim bộ  
5388  \n\nMột ngày nọ, Kei Nagai, một học sinh cấp b...  Mùa đông  phim bộ

In [34]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5389 entries, 0 to 5388
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        5389 non-null   object
 1   link         5389 non-null   object
 2   image        5389 non-null   object
 3   episode      4146 non-null   object
 4   rating       5389 non-null   object
 5   views        5389 non-null   object
 6   quality      5347 non-null   object
 7   year         5378 non-null   object
 8   genres       5359 non-null   object
 9   description  5351 non-null   object
 10  season       5389 non-null   object
 11  type         5389 non-null   object
dtypes: object(12)
memory usage: 505.3+ KB


In [35]:
data.isnull().sum()

title             0
link              0
image             0
episode        1243
rating            0
views             0
quality          42
year             11
genres           30
description      38
season            0
type              0
dtype: int64

#### Nhận xét tổng quan

- Dữ liệu gồm 5389 bản ghi với 12 cột.
- Có 5 trường dữ liệu chứa giá trị null là `episode`, `quality`, `year`, `genres` và `description`.
- Có 12 trường dữ liệu thuộc dạng `object`.

## Data cleaning

### Xử lý các ký tự \n của Description

In [36]:
data['description'] = data['description'].str.replace('\n','')
data['views'] = data['views'].str.replace(',','')
data['title'] = data['title'].str.replace(',',' ')
data['episode'] = data['episode'].str.replace('\n',' ')

### Xử lý các giá trị null

#### episode

In [37]:
data['episode'].value_counts(dropna=False)

episode
HOÀN TẤT    3505
NaN         1243
TẬP 02        50
TẬP 12        35
TẬP 03        25
            ... 
TẬP 154        1
TẬP 1149       1
TẬP 290        1
TẬP 263        1
TẬP 73         1
Name: count, Length: 124, dtype: int64

In [38]:
data[data['episode'].isnull()]

title  \
0                                 5-toubun no Hanayome*   
1                      Overlord Movie 3: Sei Oukoku-hen   
2                   Digimon Adventure 02: The Beginning   
3        Tengen Toppa Gurren Lagann Movie 1: Gurren-hen   
4                           Yowamushi Pedal: Spare Bike   
...                                                 ...   
1238                Bức Thư Từ Thế Giới Bên Kia Special   
1239                                  Shisha no Teikoku   
1240  Trinity Seven Movie: Eternity Library to Alche...   
1241                                      Thất Tinh OVA   
1242                                          Air Movie   

                                                   link  \
0     https://animevietsub.page/phim/5-toubun-no-han...   
1     https://animevietsub.page/phim/overlord-movie-...   
2     https://animevietsub.page/phim/digimon-adventu...   
3     https://animevietsub.page/phim/tengen-toppa-gu...   
4     https://animevietsub.page/phim/yowamushi-pedal...   
...                                                 ...   
1238  https://animevietsub.page/phim/buc-thu-tu-the-...   
1239  https://animevietsub.page/phim/shisha-no-teiko...   
1240  https://animevietsub.page/phim/trinity-seven-m...   
1241  https://animevietsub.page/phim/that-tinh-ova-i...   
1242     https://animevietsub.page/phim/air-movie-a188/   

                                                  image episode  \
0     https://cdn.animevietsub.page/data/poster/2024...     NaN   
1     https://cdn.animevietsub.page/data/poster/2024...     NaN   
2     https://cdn.animevietsub.page/data/poster/2024...     NaN   
3     https://cdn.animevietsub.page/data/poster/2024...     NaN   
4     https://cdn.animevietsub.page/data/poster/2024...     NaN   
...                                                 ...     ...   
1238  https://cdn.animevietsub.page/data/poster/2017...     NaN   
1239  https://cdn.animevietsub.page/data/poster/2017...     NaN   
1240  https://cdn.animevietsub.page/data/poster/2017...     NaN   
1241  https://cdn.animevietsub.page/data/poster/2017...     NaN   
1242  https://cdn.animevietsub.page/data/poster/2017...     NaN   

                                                rating   views quality  year  \
0     9.6 trong số 10 dựa trên 301 thành viên đánh giá  235236      HD  2024   
1     9.4 trong số 10 dựa trên 115 thành viên đánh giá  687666     CAM  2024   
2       7.3 trong số 10 dựa trên 3 thành viên đánh giá   27447      HD  2023   
3         9 trong số 10 dựa trên 9 thành viên đánh giá   37285     FHD  2008   
4      7.5 trong số 10 dựa trên 11 thành viên đánh giá   15265     FHD  2016   
...                                                ...     ...     ...   ...   
1238    9.5 trong số 10 dựa trên 6 thành viên đánh giá   19154      HD  2008   
1239   9.3 trong số 10 dựa trên 14 thành viên đánh giá   48964      HD  2015   
1240  9.6 trong số 10 dựa trên 109 thành viên đánh giá  425580      BD  2017   
1241   9.6 trong số 10 dựa trên 62 thành viên đánh giá  350483      BD  2015   
1242    8.3 trong số 10 dựa trên 3 thành viên đánh giá   24147      HD  2005   

                                                 genres  \
0               Shounen, School, Romance, Harem, Comedy   
1                     Fantasy, Adventure, Magic, Action   
2             Fantasy, Drama, Comedy, Adventure, Action   
3                                        Sci-Fi, Action   
4                        Shounen, Sports, Drama, Action   
...                                                 ...   
1238            Thriller, Psychological, Fantasy, Drama   
1239                  Sci-Fi, Psychological, Historical   
1240  Shounen, School, Romance, Harem, Fantasy, Ecch...   
1241  School, Romance, Supernatural, Harem, Fantasy,...   
1242                       Romance, Supernatural, Drama   

                                            description    season     type  
0     Chuyến đi tuần trăng mật của Fuutarou Uesugi v...   Mùa thu  phim lẻ  
1  

Dựa vào thống kê giá trị của `episode` và xem các giá trị null của `episode`, ta thấy tất cả các giá trị NaN đều là thuộc loại phim lẻ. Nhưng đối với loại phim lẻ thì chỉ là 1 tập vì nếu có phần 2 thì sẽ gọi bằng tên gọi khác nên sẽ thay thế các giá trị NaN bằng 1

In [39]:
data['episode'] = data['episode'].fillna(1)

#### quality

In [40]:
data['quality'].value_counts(dropna=False)

quality
HD                4397
FHD                364
BD                 282
SD                 231
NaN                 42
BD/Bluray           36
HQ                  23
CAM                  2
Full HD              2
BD Vietsub           1
480P                 1
OVA BD Vietsub       1
CAM HD               1
BD FHD               1
Bluray Disc          1
undefined            1
rHD                  1
hd                   1
26/10/2018           1
Name: count, dtype: int64

có các độ phân giải chưa phù hợp và gọi khác nhau nên sẽ tiếng hành thay thế

In [41]:
# Dựng từ điển để gộp các giá trị tương tự vào nhóm chung
quality_mapping = {
    'HD': 'HD', 'FHD': 'Full HD', 'Full HD': 'Full HD', 'BD': 'BD', 
    'BD/Bluray': 'BD', 'HQ': 'HQ', 'SD': 'SD', 'CAM': 'CAM', 
    'BD Vietsub': 'BD', '480P': 'SD', 'OVA BD Vietsub': 'BD', 
    'CAM HD': 'CAM', 'BD FHD': 'Full HD', 'Bluray Disc': 'BD', 
    'rHD': 'HD', 'hd': 'HD', 'NaN': 'Unknown', 'undefined': 'Unknown',
    '26/10/2018': 'Unknown'
}

# Áp dụng từ điển để thay thế các giá trị trong cột 'quality'
data['quality'] = data['quality'].replace(quality_mapping)
data['quality'] = data['quality'].fillna('Unknown')

# Tính lại thống kê số lượng các giá trị trong cột 'quality' sau khi gộp
quality_stats = data['quality'].value_counts(dropna=False)

# In kết quả thống kê
print(quality_stats)

quality
HD         4399
Full HD     367
BD          321
SD          232
Unknown      44
HQ           23
CAM           3
Name: count, dtype: int64


#### year

In [42]:
data['year'].value_counts(dropna=False)

year
2016         351
2018         344
2021         338
2023         329
2022         325
2024         305
2017         303
2014         303
2015         295
2019         288
2013         254
2020         253
2012         224
2011         189
2009         152
2010         152
2008         134
2007         116
2006         108
2005          88
2004          64
2003          51
2025          51
2002          50
2001          35
2000          34
1999          28
1998          27
1997          18
1994          18
1993          15
1995          15
1996          15
1988          14
1992          13
1989          13
NaN           11
1987          11
1984           7
1991           7
1986           6
1985           6
1990           6
1983           4
1981           4
1979           3
1982           3
1980           2
1978           2
1971           1
2099           1
1974           1
07 - 2021      1
2010-2011      1
Name: count, dtype: int64

như vậy có các giá trị chưa hợp lệ như là null, 07 - 2021 và 2010-2011, cách giải quyết như sau:<br>
1. null giải quyết bằng thay thế giá trị gần nhất phía dưới (backward fill)
2. 07 - 2021 thì thay thế bằng 2021 vì đó chỉ là sai định dạng
3. 2010-2011 thì thay thế bằng 2010 vì số lượng 2010 ít hơn 2011 nên thay vào không ảnh hưởng nhiều

In [43]:
data['year'] = data['year'].bfill()
data['year'] = data['year'].str.replace('07 - 2021','2021')
data['year'] = data['year'].str.replace('2010-2011','2010')

In [44]:
data['year'].value_counts(dropna=False)

year
2016    352
2018    344
2021    339
2022    330
2023    330
2024    305
2017    303
2014    303
2015    295
2019    288
2013    255
2020    254
2012    224
2011    189
2010    153
2009    152
2008    134
2007    116
2006    109
2005     88
2004     64
2025     51
2003     51
2002     50
2001     35
2000     34
1999     29
1998     27
1994     18
1997     18
1995     15
1996     15
1993     15
1988     14
1989     13
1992     13
1987     11
1984      7
1991      7
1986      6
1985      6
1990      6
1983      4
1981      4
1979      3
1982      3
1980      2
1978      2
1971      1
2099      1
1974      1
Name: count, dtype: int64

#### genres

In [45]:
null_count = data['genres'].isnull().sum()

print(f"Số lượng giá trị null trong cột 'genres': {null_count}")

Số lượng giá trị null trong cột 'genres': 30


Vì mục `genres` rất khó xác định được thể loại chính xác nên sẽ thay thế bằng Unknown

In [46]:
data['genres'] = data['genres'].fillna('Unknown')

#### description

Cũng như `genres`, `description` cũng không thể xác định bình thường nên sẽ quy về giá trị đồng nhất là `Đang cập nhật`

In [47]:
data['description'] = data['description'].fillna('Đang cập nhật')

In [48]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5389 entries, 0 to 5388
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   title        5389 non-null   object
 1   link         5389 non-null   object
 2   image        5389 non-null   object
 3   episode      5389 non-null   object
 4   rating       5389 non-null   object
 5   views        5389 non-null   object
 6   quality      5389 non-null   object
 7   year         5389 non-null   object
 8   genres       5389 non-null   object
 9   description  5389 non-null   object
 10  season       5389 non-null   object
 11  type         5389 non-null   object
dtypes: object(12)
memory usage: 505.3+ KB


Sau khi kiểm tra lại thì xác nhận toàn bộ dữ liệu đã được xử lý giá trị `null`

### Tạo thêm cột dữ liệu

Ta sẽ dùng cột `rating` để tạo ra cột `rate` và `nums_of_vote`

In [49]:
def extract_rate_and_votes(rating):
    match = re.match(r"([0-9.]+) trong số 10 dựa trên ([0-9,]+) thành viên đánh giá", rating)
    
    if match:
        rate = float(match.group(1))
        nums_of_vote = int(match.group(2).replace(',', ''))
        return rate, nums_of_vote
    else:
        return None, None

In [50]:
data[['rate', 'nums_of_vote']] = data['rating'].apply(lambda x: pd.Series(extract_rate_and_votes(x)))

In [51]:
data.head(1)

title                                               link  \
0  5-toubun no Hanayome*  https://animevietsub.page/phim/5-toubun-no-han...   

                                               image episode  \
0  https://cdn.animevietsub.page/data/poster/2024...       1   

                                             rating   views quality  year  \
0  9.6 trong số 10 dựa trên 301 thành viên đánh giá  235236      HD  2024   

                                    genres  \
0  Shounen, School, Romance, Harem, Comedy   

                                         description   season     type  rate  \
0  Chuyến đi tuần trăng mật của Fuutarou Uesugi v...  Mùa thu  phim lẻ   9.6   

   nums_of_vote  
0         301.0

In [52]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5389 entries, 0 to 5388
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   title         5389 non-null   object 
 1   link          5389 non-null   object 
 2   image         5389 non-null   object 
 3   episode       5389 non-null   object 
 4   rating        5389 non-null   object 
 5   views         5389 non-null   object 
 6   quality       5389 non-null   object 
 7   year          5389 non-null   object 
 8   genres        5389 non-null   object 
 9   description   5389 non-null   object 
 10  season        5389 non-null   object 
 11  type          5389 non-null   object 
 12  rate          5389 non-null   float64
 13  nums_of_vote  5389 non-null   float64
dtypes: float64(2), object(12)
memory usage: 589.5+ KB


### Xử lý kiểu dữ liệu

Các trường `views`, `nums_of_vote` là kiểu `int`

In [53]:
data['views'] = data['views'].astype(int)
data['nums_of_vote'] = data['nums_of_vote'].astype(int)

In [54]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5389 entries, 0 to 5388
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   title         5389 non-null   object 
 1   link          5389 non-null   object 
 2   image         5389 non-null   object 
 3   episode       5389 non-null   object 
 4   rating        5389 non-null   object 
 5   views         5389 non-null   int64  
 6   quality       5389 non-null   object 
 7   year          5389 non-null   object 
 8   genres        5389 non-null   object 
 9   description   5389 non-null   object 
 10  season        5389 non-null   object 
 11  type          5389 non-null   object 
 12  rate          5389 non-null   float64
 13  nums_of_vote  5389 non-null   int64  
dtypes: float64(1), int64(2), object(11)
memory usage: 589.5+ KB


In [55]:
os.makedirs(cleaned_data_path, exist_ok=True)
print(f"Thư mục '{cleaned_data_path}' đã được tạo hoặc đã tồn tại.")

Thư mục './data/cleaned/' đã được tạo hoặc đã tồn tại.


In [56]:
data.to_csv(csv_file_path, index=False)
print(f"Dữ liệu đã được lưu vào: {csv_file_path}")

Dữ liệu đã được lưu vào: ./data/cleaned/anime_data_cleaned.csv
